### PR TITLE
implement unix socket cleanup

### LINKED ISSUE #10969: Delete unix sockets on daemon shutdown
**Summary**: 

The daemon used to clean up its unix sockets after #6962 got merged, but there was a regression in #7317 and now it's not the case anymore.

**Proposed solution**:
Implement `removeUnixSockets` as a method on `Daemon` and call it during daemon shutdown.

### PR DESCRIPTION
**Pull request by cpuguy83**
Closes #10969

Carry of #11085

### CODE REVIEW

**Comment by jfrazelle** [comment]
thans so mcuh @cpuguy83

**Comment by cpuguy83** [comment]
This is not quite so straight-forward.
Since the api server need to be told to shutdown in order to cleanup, and seemingly the only way to tell it to shutdown is via an eng.OnShutdown handler.

This seems ok, but the handler for this is getting registered before the daemon shutdown handlers (really it's a race, but probably api will always get in first)... and for some reason the daemon shutdown handlers aren't running fully when the api is shutdown first.
We actually end up with the following log events:

```
INFO[1916] Received signal 'interrupt', starting shutdown of docker...
DEBU[1916] starting clean shutdown of all containers...
accept unix /var/run/docker.sock: use of closed network connection
INFO[1916] -job serveapi(unix:///var/run/docker.sock) = ERR (1)
ERRO[1916] ServeAPI error: accept unix /var/run/docker.sock: use of closed network connection
```

**Comment by tiborvass** [comment]
@cpuguy83 rebase please

**Comment by cpuguy83** [comment]
I'm going to close this because it's broken and I don't currently see a way to fix it.

**Comment by cpuguy83** [comment]
Ok, I had to add a wait for the engine shutdown to make sure the daemon doesn't exit after the api server is closed.

**Comment by cpuguy83** [comment]
ping @estesp

File: docker/daemon.go [hunk 1]
```diff
@@ -189,5 +189,13 @@ func mainDaemon() {
...
 		log.Errorf("Shutting down due to ServeAPI error: %v", errAPI)
 	}
+
 	eng.Shutdown()
+	// Can't just call eng.Shutdown() here and expect it to wait if the eng shutdown was already called from another goroutine
+	// Need to block here to make sure if this was called by another goroutinue that it actually finishes
+	eng.WaitShutdown()
+
+	if errAPI != nil {
+		os.Exit(1)
```
Comment (estesp) [inline docker/daemon.go]: should the separate checks for `errAPI != nil` be combined here at the end and just use `log.Fatalf(..)` to both output the log and exit with error?

File: api/server/server.go [hunk 2]
```diff
@@ -1578,7 +1578,16 @@ func ServeApi(job *engine.Job) engine.Status {
...
+			job.Eng.OnShutdown(func() {
+				if err := srv.Close(); err != nil {
+					log.Error(err)
+				}
+			})
+			if err = srv.Serve(); err != nil && strings.Contains(err.Error(), "use of closed network connection") {
+				err = nil
+			}
+			chErrors <- err
+			srv.Close()
```
Comment (estesp) [inline api/server/server.go]: when I looked at this the other day I was curious if we need both `srv.Close()` calls? The shutdown handler registered above will call it, and we call it directly after the return of the `srv.Serve()` call here..

File: docker/daemon.go (hunk 1 above)
Comment (cpuguy83) [inline docker/daemon.go]: Makes sense.

File: api/server/server.go (hunk 2 above)
Comment (cpuguy83) [inline api/server/server.go]: Yeah, we probably don't now that the shutdown stuff is fixed.

**Comment by cpuguy83** [comment]
Actually the engine shutdown is still racey here.
Closing again while I mess around with this.

**Comment by cpuguy83** [comment]
Accidently forced pushed while the old PR was closed, new one is #11525

**Pull request closed without merging**